File: go/tree/leetcode/binary_tree_inorder_traversal.go
package leetcode

import (
	"container/list"
)

//二叉树的中序遍历

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//递归
func inorderTraversal(root *TreeNode) []int {
	var res []int
	inHelper(root, &res)
	return res
}

func inHelper(node *TreeNode, res *[]int) {
	if node == nil {
		return
	}

	inHelper(node.Left, res)
	*res = append(*res, node.Val)
	inHelper(node.Right, res)
}

// 递归-另一种写法
func inorderTraversal2(root *TreeNode) (res []int) {
	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		res = append(res, node.Val)
		inorder(node.Right)
	}
	inorder(root)
	return
}

//使用一个栈
func inorderTraversal3(root *TreeNode) []int {
	stack := list.New()
	var res []int
	cur := root
	for cur != nil || stack.Len() > 0 {
		for cur != nil {
			stack.PushFront(cur)
			cur = cur.Left
		}

		e := stack.Front()
		res = append(res, e.Value.(*TreeNode).Val)
		stack.Remove(e)
		cur = e.Value.(*TreeNode).Right
	}
	return res
}

// 使用数组当栈
func inorderTraversal4(root *TreeNode) []int {
	var res []int
	var stack []*TreeNode
	cur := root
	for cur != nil || len(stack) > 0 {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		res = append(res, node.Val)
		cur = node.Right
	}
	return res
}

File: java/datastructure/tree/leetcode/BinaryTreeInorderTraversal.java
package datastructure.tree.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author roseduan
 * @time 2020/9/14 10:03 下午
 * @description 二叉树的中序遍历
 */
public class BinaryTreeInorderTraversal {

    private final List<Integer> res = new ArrayList<>();

    /**
     * 递归的方式
     * 运行结果：通过，速度击败 100%
     */
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return res;
        }
        inorderTraversal(root.left);
        res.add(root.val);
        inorderTraversal(root.right);
        return res;
    }

    /**
     * 迭代的方式，使用一个栈，手动实现递归的效果
     */
    public List<Integer> inorderTraversal2(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (cur != null || !stack.empty()) {
            while (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            res.add(cur.val);
            cur = cur.right;
        }
        return res;
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
